[dtensor] handle the case where output of op is Optional[Tensor]

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90241
* #90106
* #89970
* #89969
* #89968
* #89967

Observed by @aazzolini, some op might have Optional[Tensor] returns
where it return None (i.e. native_layer_norm_backward), it's a mismatch
between C++ aten op signature and python None, but we need to handle it
in the python side

wanchaol committed:
[dtensor] handle the case where output of op is Optional[Tensor]

Observed by @aazzolini, some op might have Optional[Tensor] returns
where it return None (i.e. native_layer_norm_backward), it's a mismatch
between C++ aten op signature and python None, but we need to handle it
in the python side

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] handle the case where output of op is Optional[Tensor]"


Observed by aazzolini, some op might have Optional[Tensor] returns
where it return None (i.e. native_layer_norm_backward), it's a mismatch
between C++ aten op signature and python None, but we need to handle it
in the python side

[ghstack-poisoned]

aazzolini reviewed (commented):
Typing wise you'll also have to modify the signature for OutputSharding as well. Gotta return a Sequence[Optional[DTensorSpec]]  (instead of Sequence[DTensorSpec])

wanchaol committed:
Update on "[dtensor] handle the case where output of op is Optional[Tensor]"


Observed by aazzolini, some op might have Optional[Tensor] returns
where it return None (i.e. native_layer_norm_backward), it's a mismatch
between C++ aten op signature and python None, but we need to handle it
in the python side

[ghstack-poisoned]

aazzolini reviewed (approved):
Thanks, @wanchaol !

wanchaol commented:
@pytorchbot merge -f "failure not related"
